packages/api-fetch/src/index.js

```diff
@@ -156,6 +157,42 @@ function apiFetch( options ) {
 	} );
 }
 
+/**
+ * Function that fetches data using apiFetch, and updates the status.
+ *
+ * @param {string} path Query path.
+ */
+function useApiFetch( path ) {
+	// Indicate the fetching status
+	const [ isLoading, setIsLoading ] = useState( true );
+	const [ data, setData ] = useState( null );
+	const [ error, setError ] = useState( null );
+
+	useEffect( () => {
+		setIsLoading( true );
+		setData( null );
+		setError( null );
+
+		apiFetch( { path } )
+			.then( ( fetchedData ) => {
+				setData( fetchedData );
+				// We've stopped fetching
+				setIsLoading( false );
+			} )
+			.catch( ( err ) => {
+				setError( err );
+				// We've stopped fetching
+				setIsLoading( false );
+			} );
```

Comment:
There's no guarantee that the component is still mounted at the point we're calling `setState` here. If it was removed in the time between the request starting and its completion, the state setters will fail, and React will log a warning to the console.

See also: https://reactjs.org/docs/hooks-reference.html#cleaning-up-an-effect

Reply:
Can I also reinforce that this anti pattern is super easy to introduce. I wonder whether we could make the hook handle this edge cases automatically by:

1. Wrapping the fetch Promise to allow to be cancelled.
2. Automatically calling the `.cancel()` on unmount.
3. Checking for `isCancelled` state in `.then` / `.catch` handlers and exiting early ( to avoid state updates).

For an example see https://github.com/Automattic/jetpack/pull/15228

Reply:
This point makes me think of the behaviour of resolvers in `wp.data` which often include making fetch requests. This same issue can surface there (and I think [already is](https://github.com/WordPress/gutenberg/issues/21049)). We probably should be looking into that too.

Reply:
Maybe we could promote good behaviour here by having `apiFetch` return a function that cancels the request.

```js
useEffect( () => {
	setIsLoading( true );
	setData( null );
	setError( null );

	const cancelFetch = apiFetch( { path } )
		.then( ( fetchedData ) => {
			setData( fetchedData );
			// We've stopped fetching
			setIsLoading( false );
		} )
		.catch( ( err ) => {
			setError( err );
			// We've stopped fetching
			setIsLoading( false );
		} );

	return cancelFetch;
}, [ path ] );
```

Reply:
I'm motivated toward anything which keeps `apiFetch` cumbersome to use in a component, if it means that the preferred core data entities are the more obvious solution by virtue of their ergonomics advantage. Which isn't to say there's not issues there as well (https://github.com/WordPress/gutenberg/pull/18669#discussion_r402223905), but I think we're in a better position to solve those internal to those implementations, even if it's [something like this](https://github.com/WordPress/gutenberg/blob/63a2e776f73c6cc626bc827802341c1178ee2814/packages/block-editor/src/components/link-control/index.js#L55-L80).